File: HomeKitMCP/Utilities/CharacteristicTypes.swift
import HomeKit

/// Maps HomeKit characteristic type UUIDs to human-readable names.
enum CharacteristicTypes {
    private static let mapping: [String: String] = [
        HMCharacteristicTypePowerState: "Power",
        HMCharacteristicTypeBrightness: "Brightness",
        HMCharacteristicTypeHue: "Hue",
        HMCharacteristicTypeSaturation: "Saturation",
        HMCharacteristicTypeColorTemperature: "Color Temperature",
        HMCharacteristicTypeCurrentTemperature: "Current Temperature",
        HMCharacteristicTypeTargetTemperature: "Target Temperature",
        HMCharacteristicTypeTemperatureUnits: "Temperature Units",
        HMCharacteristicTypeCurrentHeatingCooling: "Current Mode",
        HMCharacteristicTypeTargetHeatingCooling: "Target Mode",
        HMCharacteristicTypeCurrentRelativeHumidity: "Current Humidity",
        HMCharacteristicTypeTargetRelativeHumidity: "Target Humidity",
        HMCharacteristicTypeCurrentDoorState: "Door State",
        HMCharacteristicTypeTargetDoorState: "Target Door State",
        HMCharacteristicTypeCurrentLockMechanismState: "Lock State",
        HMCharacteristicTypeTargetLockMechanismState: "Target Lock State",
        HMCharacteristicTypeCurrentPosition: "Current Position",
        HMCharacteristicTypeTargetPosition: "Target Position",
        HMCharacteristicTypePositionState: "Position State",
        HMCharacteristicTypeMotionDetected: "Motion Detected",
        HMCharacteristicTypeContactState: "Contact State",
        HMCharacteristicTypeOccupancyDetected: "Occupancy Detected",
        HMCharacteristicTypeSmokeDetected: "Smoke Detected",
        HMCharacteristicTypeCarbonMonoxideDetected: "CO Detected",
        HMCharacteristicTypeBatteryLevel: "Battery Level",
        HMCharacteristicTypeStatusLowBattery: "Low Battery",
        HMCharacteristicTypeChargingState: "Charging State",
        HMCharacteristicTypeOutletInUse: "Outlet In Use",
        HMCharacteristicTypeRotationSpeed: "Rotation Speed",
        HMCharacteristicTypeCurrentFanState: "Fan State",
        HMCharacteristicTypeTargetFanState: "Target Fan State",
        HMCharacteristicTypeActive: "Active",
        HMCharacteristicTypeStatusActive: "Status Active",
        HMCharacteristicTypeName: "Name",
        HMCharacteristicTypeObstructionDetected: "Obstruction Detected",
        HMCharacteristicTypeStatusFault: "Fault",
        HMCharacteristicTypeStatusTampered: "Tampered",
        HMCharacteristicTypeCurrentLightLevel: "Light Level",
        HMCharacteristicTypeInputEvent: "Input Event",
        HMCharacteristicTypeProgramMode: "Program Mode",
        HMCharacteristicTypeInUse: "In Use",
        HMCharacteristicTypeRemainingDuration: "Remaining Duration",
        HMCharacteristicTypeSetDuration: "Set Duration",
        HMCharacteristicTypeValveType: "Valve Type",
        HMCharacteristicTypeIsConfigured: "Is Configured",
    ]

    /// Semantic descriptions explaining what each characteristic represents.
    private static let descriptions: [String: String] = [
        HMCharacteristicTypePowerState: "On/off state of a device",
        HMCharacteristicTypeBrightness: "Light intensity level",
        HMCharacteristicTypeHue: "Color hue component on the color wheel",
        HMCharacteristicTypeSaturation: "Color saturation intensity; 0% is white, 100% is full color",
        HMCharacteristicTypeColorTemperature: "White light warmth; low values are warm (yellow), high values are cool (blue)",
        HMCharacteristicTypeCurrentTemperature: "Ambient temperature reading from a sensor (read-only)",
        HMCharacteristicTypeTargetTemperature: "Desired temperature setpoint for a thermostat",
        HMCharacteristicTypeTemperatureUnits: "Display unit preference for temperature (Celsius or Fahrenheit)",
        HMCharacteristicTypeCurrentHeatingCooling: "Current operating mode of a thermostat (off, heating, or cooling) (read-only)",
        HMCharacteristicTypeTargetHeatingCooling: "Desired operating mode for a thermostat (off, heat, cool, or auto)",
        HMCharacteristicTypeCurrentRelativeHumidity: "Current relative humidity reading from a sensor (read-only)",
        HMCharacteristicTypeTargetRelativeHumidity: "Desired humidity level for a humidifier/dehumidifier",
        HMCharacteristicTypeCurrentDoorState: "Current physical state of a door (open, closed, opening, closing, stopped) (read-only)",
        HMCharacteristicTypeTargetDoorState: "Desired state for a motorized door (open or closed)",
        HMCharacteristicTypeCurrentLockMechanismState: "Current state of a lock (secured, unsecured, jammed, unknown) (read-only)",
        HMCharacteristicTypeTargetLockMechanismState: "Desired state for a lock (secured or unsecured)",
        HMCharacteristicTypeCurrentPosition: "Current position of a window covering, door, or window as percentage (read-only)",
        HMCharacteristicTypeTargetPosition: "Desired position for a window covering, door, or window as percentage",
        HMCharacteristicTypePositionState: "Movement direction of a positionable device (increasing, decreasing, stopped) (read-only)",
        HMCharacteristicTypeMotionDetected: "Whether motion is currently detected in the sensor's field (read-only)",
        HMCharacteristicTypeContactState: "Whether a door or window is open or closed (read-only)",
        HMCharacteristicTypeOccupancyDetected: "Whether a room is currently occupied (read-only)",
        HMCharacteristicTypeSmokeDetected: "Whether smoke is detected by the sensor (read-only)",
        HMCharacteristicTypeCarbonMonoxideDetected: "Whether carbon monoxide is at abnormal levels (read-only)",
        HMCharacteristicTypeBatteryLevel: "Remaining battery charge as a percentage (read-only)",
        HMCharacteristicTypeStatusLowBattery: "Whether battery level is critically low (read-only)",
        HMCharacteristicTypeChargingState: "Whether the device battery is currently charging (read-only)",
        HMCharacteristicTypeOutletInUse: "Whether an outlet is currently supplying power to a connected device (read-only)",
        HMCharacteristicTypeRotationSpeed: "Fan or motor rotation speed as a percentage",
        HMCharacteristicTypeCurrentFanState: "Current operating state of a fan (inactive, idle, blowing) (read-only)",
        HMCharacteristicTypeTargetFanState: "Desired fan operating mode (manual or auto)",
        HMCharacteristicTypeActive: "Whether a device is actively operating (e.g. air purifier running, valve open)",
        HMCharacteristicTypeStatusActive: "Whether a device is currently active and functional (read-only)",
        HMCharacteristicTypeName: "User-assigned name of the accessory or service (read-only)",
        HMCharacteristicTypeObstructionDetected: "Whether a physical obstruction is blocking a device like a garage door (read-only)",
        HMCharacteristicTypeStatusFault: "Whether the device has detected an internal fault (read-only)",
        HMCharacteristicTypeStatusTampered: "Whether the device has been physically tampered with (read-only)",
        HMCharacteristicTypeCurrentLightLevel: "Ambient light level measured in lux (read-only)",
        HMCharacteristicTypeInputEvent: "Button press event type (single press, double press, long press) (read-only)",
        HMCharacteristicTypeProgramMode: "Current scheduling/program mode of a valve or irrigation system (read-only)",
        HMCharacteristicTypeInUse: "Whether a valve or sprinkler is currently running water (read-only)",
        HMCharacteristicTypeRemainingDuration: "Seconds remaining in the current valve/sprinkler cycle (read-only)",
        HMCharacteristicTypeSetDuration: "Duration in seconds to run a valve or sprinkler when activated",
        HMCharacteristicTypeValveType: "Type of valve (generic, irrigation, shower head, faucet) (read-only)",
        HMCharacteristicTypeIsConfigured: "Whether a service has been set up and is ready to use (read-only)",
    ]

    static func displayName(for characteristicType: String) -> String {
        return mapping[characteristicType] ?? characteristicType
    }

    /// Returns the semantic description for a characteristic type, or nil if unknown.
    static func description(for characteristicType: String) -> String? {
        return descriptions[characteristicType]
    }
    
    /// All known characteristic type friendly names, sorted alphabetically.
    static var allDisplayNames: [String] {
        Array(Set(mapping.values)).sorted()
    }

    /// Returns the full mapping from HomeKit UUID to friendly name (for metadata tools).
    static var allMappings: [(uuid: String, displayName: String)] {
        mapping.map { (uuid: $0.key, displayName: $0.value) }.sorted { $0.displayName < $1.displayName }
    }

    /// Returns all aliases for a given HomeKit characteristic type UUID.
    static func aliases(for characteristicType: String) -> [String] {
        reverseMapping.compactMap { key, value in
            value == characteristicType ? key : nil
        }.sorted()
    }

    static func isSupported(_ type: String) -> Bool {
        // Strict allowlist: only characteristics with a mapped friendly name are supported.
        // This filters out vendor-specific (UUIDs) and other unsupported characteristics.
        return mapping[type] != nil
    }

    /// Maps human-readable names back to HMCharacteristic type UUIDs for MCP tool usage.
    private static let reverseMapping: [String: String] = {
        var result: [String: String] = [:]
        for (key, value) in mapping {
            result[value.lowercased()] = key
        }
        // Also add common shorthand names
        result["power"] = HMCharacteristicTypePowerState
        result["brightness"] = HMCharacteristicTypeBrightness
        result["hue"] = HMCharacteristicTypeHue
        result["saturation"] = HMCharacteristicTypeSaturation
        result["temperature"] = HMCharacteristicTypeTargetTemperature
        result["current_temperature"] = HMCharacteristicTypeCurrentTemperature
        result["target_position"] = HMCharacteristicTypeTargetPosition
        result["lock_state"] = HMCharacteristicTypeTargetLockMechanismState
        result["color_temperature"] = HMCharacteristicTypeColorTemperature
        result["rotation_speed"] = HMCharacteristicTypeRotationSpeed
        return result
    }()

    static func characteristicType(forName name: String) -> String? {
        return reverseMapping[name.lowercased()]
    }

    static func formatValue(_ value: Any, characteristicType: String) -> String {
        switch characteristicType {
        case HMCharacteristicTypePowerState,
             HMCharacteristicTypeMotionDetected,
             HMCharacteristicTypeContactState,
             HMCharacteristicTypeOccupancyDetected,
             HMCharacteristicTypeSmokeDetected,
             HMCharacteristicTypeCarbonMonoxideDetected,
             HMCharacteristicTypeOutletInUse,
             HMCharacteristicTypeObstructionDetected,
             HMCharacteristicTypeStatusActive,
             HMCharacteristicTypeActive:
            if let boolVal = value as? Bool {
                return boolVal ? "On" : "Off"
            }
            if let intVal = value as? Int {
                return intVal != 0 ? "On" : "Off"
            }

        case HMCharacteristicTypeBrightness,
             HMCharacteristicTypeSaturation,
             HMCharacteristicTypeBatteryLevel,
             HMCharacteristicTypeCurrentRelativeHumidity,
             HMCharacteristicTypeTargetRelativeHumidity:
            return "\(value)%"

        case HMCharacteristicTypeCurrentTemperature,
             HMCharacteristicTypeTargetTemperature:
            if let doubleVal = value as? Double {
                return String(format: "%.1f%@", doubleVal, TemperatureConversion.unitSuffix)
            }

        case HMCharacteristicTypeHue:
            return "\(value)°"

        case HMCharacteristicTypeCurrentDoorState:
            if let intVal = value as? Int {
                switch intVal {
                case 0: return "Open"
                case 1: return "Closed"
                case 2: return "Opening"
                case 3: return "Closing"
                case 4: return "Stopped"
                default: break
                }
            }

        case HMCharacteristicTypeCurrentLockMechanismState:
            if let intVal = value as? Int {
                switch intVal {
                case 0: return "Unsecured"
                case 1: return "Secured"
                case 2: return "Jammed"
                case 3: return "Unknown"
                default: break
                }
            }

        case HMCharacteristicTypeCurrentPosition,
             HMCharacteristicTypeTargetPosition:
            return "\(value)%"

        case HMCharacteristicTypeColorTemperature:
            return "\(value)K"

        case HMCharacteristicTypeRotationSpeed:
            return "\(value)%"

        default:
            break
        }

        return "\(value)"
    }
}

File: log-viewer-web/src/utils/characteristic-types.ts
const CHARACTERISTIC_DISPLAY_NAMES: Record<string, string> = {
  '00000025-0000-1000-8000-0026BB765291': 'Power',
  '00000008-0000-1000-8000-0026BB765291': 'Brightness',
  '00000013-0000-1000-8000-0026BB765291': 'Hue',
  '0000002F-0000-1000-8000-0026BB765291': 'Saturation',
  '000000CE-0000-1000-8000-0026BB765291': 'Color Temperature',
  '00000011-0000-1000-8000-0026BB765291': 'Current Temperature',
  '00000035-0000-1000-8000-0026BB765291': 'Target Temperature',
  '00000036-0000-1000-8000-0026BB765291': 'Temperature Units',
  '0000000F-0000-1000-8000-0026BB765291': 'Current Mode',
  '00000033-0000-1000-8000-0026BB765291': 'Target Mode',
  '00000010-0000-1000-8000-0026BB765291': 'Current Humidity',
  '00000034-0000-1000-8000-0026BB765291': 'Target Humidity',
  '0000000E-0000-1000-8000-0026BB765291': 'Door State',
  '00000032-0000-1000-8000-0026BB765291': 'Target Door State',
  '0000001D-0000-1000-8000-0026BB765291': 'Lock State',
  '0000001E-0000-1000-8000-0026BB765291': 'Target Lock State',
  '0000006D-0000-1000-8000-0026BB765291': 'Current Position',
  '0000007C-0000-1000-8000-0026BB765291': 'Target Position',
  '00000072-0000-1000-8000-0026BB765291': 'Position State',
  '00000022-0000-1000-8000-0026BB765291': 'Motion Detected',
  '0000006A-0000-1000-8000-0026BB765291': 'Contact State',
  '00000071-0000-1000-8000-0026BB765291': 'Occupancy Detected',
  '00000076-0000-1000-8000-0026BB765291': 'Smoke Detected',
  '00000069-0000-1000-8000-0026BB765291': 'CO Detected',
  '00000068-0000-1000-8000-0026BB765291': 'Battery Level',
  '00000079-0000-1000-8000-0026BB765291': 'Low Battery',
  '0000008F-0000-1000-8000-0026BB765291': 'Charging State',
  '00000026-0000-1000-8000-0026BB765291': 'Outlet In Use',
  '00000029-0000-1000-8000-0026BB765291': 'Rotation Speed',
  '000000AF-0000-1000-8000-0026BB765291': 'Fan State',
  '000000BF-0000-1000-8000-0026BB765291': 'Target Fan State',
  '000000B0-0000-1000-8000-0026BB765291': 'Active',
  '00000075-0000-1000-8000-0026BB765291': 'Status Active',
  '00000023-0000-1000-8000-0026BB765291': 'Name',
  '00000024-0000-1000-8000-0026BB765291': 'Obstruction Detected',
  '00000077-0000-1000-8000-0026BB765291': 'Fault',
  '0000007A-0000-1000-8000-0026BB765291': 'Tampered',
  '0000006B-0000-1000-8000-0026BB765291': 'Light Level',
  '00000073-0000-1000-8000-0026BB765291': 'Input Event',
  '000000D1-0000-1000-8000-0026BB765291': 'Program Mode',
  '000000D2-0000-1000-8000-0026BB765291': 'In Use',
  '000000D4-0000-1000-8000-0026BB765291': 'Remaining Duration',
  '000000D3-0000-1000-8000-0026BB765291': 'Set Duration',
  '000000D5-0000-1000-8000-0026BB765291': 'Valve Type',
  '000000D6-0000-1000-8000-0026BB765291': 'Is Configured',
};

const BOOLEAN_TYPES = new Set([
  'Power', 'Motion Detected', 'Contact State', 'Occupancy Detected',
  'Smoke Detected', 'CO Detected', 'Outlet In Use', 'Obstruction Detected',
  'Status Active', 'Active',
]);

const PERCENTAGE_TYPES = new Set([
  'Brightness', 'Saturation', 'Battery Level',
  'Current Humidity', 'Target Humidity',
  'Current Position', 'Target Position', 'Rotation Speed',
]);

const TEMPERATURE_TYPES = new Set(['Current Temperature', 'Target Temperature']);

const DOOR_STATES = ['Open', 'Closed', 'Opening', 'Closing', 'Stopped'];
const LOCK_STATES = ['Unsecured', 'Secured', 'Jammed', 'Unknown'];

export function characteristicDisplayName(type: string): string {
  const upper = type.toUpperCase();
  if (CHARACTERISTIC_DISPLAY_NAMES[upper]) {
    return CHARACTERISTIC_DISPLAY_NAMES[upper];
  }
  for (const [key, value] of Object.entries(CHARACTERISTIC_DISPLAY_NAMES)) {
    if (key.toUpperCase() === upper) return value;
  }
  if (!/^[0-9A-F]{8}-[0-9A-F]{4}-/i.test(type)) {
    return type
      .replace(/[-_.]/g, ' ')
      .replace(/\b\w/g, (c) => c.toUpperCase())
      .trim();
  }
  return type;
}

export function formatCharacteristicValue(value: unknown, characteristicType: string): string {
  if (value === undefined || value === null) return '--';

  const name = characteristicDisplayName(characteristicType);

  if (BOOLEAN_TYPES.has(name)) {
    if (typeof value === 'boolean') return value ? 'On' : 'Off';
    if (typeof value === 'number') return value !== 0 ? 'On' : 'Off';
  }

  if (PERCENTAGE_TYPES.has(name)) {
    return `${value}%`;
  }

  if (TEMPERATURE_TYPES.has(name)) {
    if (typeof value === 'number') return `${value.toFixed(1)}\u00B0C`;
  }

  if (name === 'Hue') return `${value}\u00B0`;
  if (name === 'Color Temperature') return `${value}K`;

  if (name === 'Door State' && typeof value === 'number') {
    return DOOR_STATES[value] ?? String(value);
  }

  if (name === 'Lock State' && typeof value === 'number') {
    return LOCK_STATES[value] ?? String(value);
  }

  if (typeof value === 'boolean') return value ? 'On' : 'Off';
  return String(value);
}
